### ISSUE TITLE
Synchronization and signal in executor::execute and related functions

### ISSUE THREAD

**Issue by LeeHowes** [executors/futures#3]
One of the actions that comes out of https://github.com/executors/futures/issues/2 is that either on setting the value (if the callback is present) or setting the callback (if the value is present) the continuation will have to be passed to an executor. There therefore needs to be synchronization on the execute functions that will both:
 * Safely enqueue the work from one TOE into an executor running on some other TOE, independent of forward progress guarantees of the TsOE, though not independent of the basic guarantee the execute operations provide.
 * Signal the executor so that if it is running on a weak agent, it is notified to start, and if is blocked on a condition variable then that signals the thread.

What is the basic guarantee we need to make this work across all TsOE?

**Comment by dhollman** [comment]
Is this just some variant of `require(ex, never_blocking).twoway_execute()`? Or am I not understanding the point here?

**Comment by LeeHowes** [comment]
Yes and no. never_blocking is a minimum in the general case (though you can imagine, for example, a fiber-aware executor allowing blocking enqueue) but exactly what it means to not block is important. Does not blocking actually mean lock-free? That's not the same thing. Maybe it should be, though.

**Comment by dhollman** [comment]
I would expect that at minimum, `never_blocking` provides some guarantee that it's going to use locks in a responsible way. It can't spin on a lock indefinitely or acquire multiple locks without some sort of deadlock avoidance algorithm — both of these are firmly in the camp of blocking behavior AFAICT. Do we actually need lock-free here? I suppose if someone is going to try and enqueue work from a signal handler or something you'd need a stronger guarantee, but maybe this should be a separate value of the blocking property for executors?

**Comment by dhollman** [comment]
I would argue that an executor with weak agents that that locks to dequeue work cannot provide (a standards-compliant version of) the `never_blocking` property. I hope I'm not in the minority on this matter, but I totally could be.

**Comment by LeeHowes** [comment]
This has significant effects on #29

**Comment by brycelelbach** [comment]
We did not get to this at the NVIDIA F2F meeting. We need some story here for V1.0.
